src/qibo/models/circuit.py

```diff
@@ -88,24 +85,27 @@ def from_fused(self):
                 queue.append(gate.gates[0])
         return queue
 
-    def append(self, gate: gates.Gate):
-        super().append(gate)
-        if gate.qubits:
-            qubits = gate.qubits
-        else:  # special gate acting on all qubits
-            qubits = tuple(range(self.nqubits))
-
-        if isinstance(gate, gates.M):
-            self.nmeasurements += 1
+    def nmeasurements(self):
+        return len([1 for gate in self if isinstance(gate, gates.M)])
 
-        # calculate moment index for this gate
-        idx = max(self.moment_index[q] for q in qubits)
-        for q in qubits:
-            if idx >= len(self.moments):
-                # Add a moment
-                self.moments.append(len(self.moments[-1]) * [None])
-            self.moments[idx][q] = gate
-            self.moment_index[q] = idx + 1
+    def moments(self):
```

Comment:
Thanks @BrunoLiegiBastonLiegi. Maybe we could define this as a `@property` since it doesn't have any arguments and also to keep the same interface as before. 

In this particular case we could even give a try to [`@cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) which could speed things up in cases where calculation of moments is slow and the property is called multiple times (however I have not really tested that). Another thing to check in that case is how it behaves if another gate is added after the cached property is called, if the cache is updated or no (I am also not sure).

Reply:
Yes I thought about the `cached_property` too, but I tested it and it doesn't work if the queue gets updated.